packet03-redirecting/README.org

```diff
@@ -5,49 +5,277 @@
+XDP_PASS               0 pkts (         0 pps)           0 Kbytes (     0 Mbits/s) period:0.250234
+XDP_TX                 0 pkts (         0 pps)           0 Kbytes (     0 Mbits/s) period:0.250231
+XDP_REDIRECT        1303 pkts (         0 pps)         153 Kbytes (     0 Mbits/s) period:0.250228
+
+^C
+$ sudo ./xdp_stats -d left
+
+Collecting stats from BPF map
+ - BPF map (bpf_map_type:6) id:186 name:xdp_stats_map key_size:4 value_size:16 max_entries:5
+XDP-action
+XDP_ABORTED            0 pkts (         0 pps)           0 Kbytes (     0 Mbits/s) period:0.250154
+XDP_DROP               0 pkts (         0 pps)           0 Kbytes (     0 Mbits/s) period:0.250206
+XDP_PASS               0 pkts (         0 pps)           0 Kbytes (     0 Mbits/s) period:0.250206
+XDP_TX                 0 pkts (         0 pps)           0 Kbytes (     0 Mbits/s) period:0.250206
+XDP_REDIRECT          22 pkts (         0 pps)           2 Kbytes (     0 Mbits/s) period:0.250206
+
+^C
+#+end_src
+
+If, however, we try to ping outer interfaces from inner or vice versa, we
+won't see any replies, as packets destined to outer interfaces will also be
+redirected. Besides that, our implementation doesn't easily scale to more
+than two interfaces. The next assignment will show how to forward packets in
+a better manner using a kernel helper.
 
 ** Assignment 4: Use the BPF helper for routing
-After completing Assignment 3, you'll have hard-coded a redirect between two
-interfaces. But wouldn't it be better to dynamically lookup where each
-packet should go? Fortunately, the kernel already has this functionality, in
-the form of a routing table. There's a BPF helper that allows XDP programs
-to do lookups in the kernel routing table, and use this information to
-redirect packets. In this assignment, extend your forwarding program to use
-this facility for full redirection.
 
+After completing Assignment 3, you'll have a hard-coded redirect between the
+two inner interfaces. As was mentioned above, we are able to deliver packets
+only between inner interfaces—a packet destined to an outer interface will be
+delivered to the opposite inner interface and dropped there because of the
+wrong destination L3 address. We can manually check the IP/IPv6 addresses
+and return =XDP_PASS= when packets destined to outer interfaces, but this
+doesn't cover all cases and wouldn't it be better to dynamically lookup
+where each packet should go?
+
+This assignment teaches to use the =bpf_fib_lookup= helper. For the most
+part, reproduces the [[https://github.com/torvalds/linux/blob/master/samples/bpf/xdp_fwd_kern.c][xdp_fwd_kern.c]]
+example from the Linux kernel, but was patched to update statistics as in other
+examples and to check all the return values from the =bpf_fib_lookup()= function.
+Don't forget to update the =xdp_prog_user.c= program to setup 1-1 mapping
+between virtual ports and network devices.
```

Comment:
This should probably be explained in a bit more detail. Maybe expand the explanation of bpf_fib_lookup() in the previous section, to explain that it returns ifindexes, but needs virtual ports?

Reply:
ok